mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxMutabilityHelper.kt

```diff
@@ -41,3 +44,31 @@ internal fun KClass<*>.assertImmutability() {
         }?.let { throw IllegalArgumentException(it) }
     }
 }
+
+/**
+ * Checks that a state's value is not changed over its lifetime.
+ */
+internal class MutableStateChecker<S : MvRxState>(val initialState: S) {
+
+    data class StateWrapper<S : MvRxState>(val state: S) {
+        private val originalHashCode = hashCode()
+
+        fun validate() = require(originalHashCode == hashCode()) {
```

Comment:
looks like every state hash is computed twice

Reply:
yes, it has to be. once when it is first set and once when the next state is set, then we can compare the hashcodes to see if they have changed.

Reply:
ic. i think the implementation can be simplified as just 2 lines of code: 

```
            stateStore.set {
                val oldHash = hash() // add this
                val firstState = reducer()
                val secondState = reducer()
                 if (firstState != secondState) throw IllegalArgumentException("Your reducer must be pure!")
                if (oldHash != hash()) throw IllegalArgumentException("You should not mutate the state")  // add this
                 firstState
            }
```

And in this way we don't have to worry about lateinit the MutableStateChecker in the comment below.

Reply:
ah, you are right!